### ISSUE TITLE
Too many pings to galley, connection closed

### ISSUE THREAD

**Issue by Stono** [istio/istio#15088]
**Bug description**
I'm seeing connections being closed in galley due to too many pings from the client:

```
2019-06-24T10:51:05.928628Z	info	transport: Got too many pings from the client, closing the connection.
2019-06-24T10:51:05.928770Z	info	transport: loopyWriter.run returning. Err: transport: Connection closing
2019-06-24T10:51:05.928866Z	info	transport: http2Server.HandleStreams failed to read frame: read tcp 10.202.7.115:9901->10.202.4.70:60468: use of closed network connection
2019-06-24T10:51:05.929078Z	error	mcp	MCP: connection {addr=10.202.4.70:60468 id=1}: TERMINATED with errors: rpc error: code = Canceled desc = context canceled
2019-06-24T10:51:05.929175Z	info	mcp	MCP: connection {addr=10.202.4.70:60468 id=1}: CLOSED
```

This ip correlates to a pilot instance:
```
istio-pilot-847cddc494-84jql              2/2     Running   0          120m    10.202.4.70    gke-delivery-platform-core-20190401-3a7a2d31-qlfk     <none>           <none>
```

And discovery logs:
```
2019-06-24T10:51:05.929040Z	info	Client received GoAway with http2.ErrCodeEnhanceYourCalm.
2019-06-24T10:51:05.929226Z	info	transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-06-24T10:51:05.929301Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc420524080, TRANSIENT_FAILURE
2019-06-24T10:51:05.929347Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc420524080, CONNECTING
2019-06-24T10:51:05.929299Z	error	mcp	Error receiving MCP resource: rpc error: code = Unavailable desc = transport is closing
2019-06-24T10:51:05.929391Z	error	mcp	Error receiving MCP response: rpc error: code = Unavailable desc = transport is closing
2019-06-24T10:51:05.944696Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc420524080, READY
```

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[x] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastrcture

**Expected behavior**
I don't know how much of an issue this is, but I am guessing it probably shouldn't be disconnecting.

**Steps to reproduce the bug**
This isn't a particularly large cluster, nor does it have particularly high churn.  `158` services with `197` pods.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
Istio 1.2.0 on GKE 1.13.6-gke.13.

**How was Istio installed?**
Helm

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

**Comment by howardjohn** [comment]
Some more info: https://github.com/istio/istio/issues/13910

@sdake @ozevren

**Comment by Stono** [comment]
For more info, we've seen this on 3 clusters all of drastically different sizes.  It only appears to happen after galley is started.

At that same time, we also have validation hook failures (see https://github.com/istio/istio/issues/15091).

**Comment by duderino** [comment]
@nmittler @ozevren I'm not sure if this is a Pilot or a Galley bug.  Could you root cause it and then assign the bug to the appropriate vict^H^H^H^Howner?

**Comment by ozevren** [comment]
This is mainly an issue with the gRPC layer. I've made some gRPC connection related knob/lever changes some time ago to alleviate this to some extent but apparently not eliminated it.

**Comment by ozevren** [comment]
I don't think this will surface as a functionality problem, but a potential performance problem. (@Stono are you seeing any abnormalities due to / correlated to this?)

**Comment by Stono** [comment]
@ozevren not so far (that I can tell).  I've been adding `1.2.0` to various non-production clusters of different sizes.  Only time will tell if an issue manifests 🤷‍♂

**Comment by ozevren** [comment]
Fixed is in for master.

**Comment by Stono** [comment]
Nice one @ozevren - I'll test it out when it's in a patch release and let you know how I get on.

**Comment by hzxuzhonghu** [comment]
I've seen this in 1.2.0 too
```
2019-06-28T02:04:03.612088Z     info    Client received GoAway with http2.ErrCodeEnhanceYourCalm.
2019-06-28T02:04:03.612265Z     info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2019-06-28T02:04:03.612339Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc420336070, TRANSIENT_FAILURE
2019-06-28T02:04:03.612427Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc420336070, CONNECTING
2019-06-28T02:04:03.612479Z     error   mcp     Error receiving MCP resource: rpc error: code = Unavailable desc = transport is closing
2019-06-28T02:04:03.612536Z     error   mcp     Error receiving MCP response: rpc error: code = Unavailable desc = transport is closing
2019-06-28T02:04:03.615016Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc420336070, READY
2019-06-28T02:04:04.612739Z     info    mcp     (re)trying to establish new MCP sink stream
2019-06-28T02:04:04.612960Z     info    mcp     New MCP sink stream created

```

**Comment by ozevren** [comment]
@hzxuzhonghu Can you check with master HEAD or 1.2.x branch HEAD?

**Comment by hzxuzhonghu** [comment]
I am gonna upgrade galley image.

**Comment by hzxuzhonghu** [comment]
I can still see GoAway error, but only one time.　
```
2019-06-29T02:09:27.854748Z     info    Client received GoAway with http2.ErrCodeEnhanceYourCalm.
2019-06-29T02:09:27.855072Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc4206a0c70, TRANSIENT_FAILURE
2019-06-29T02:09:27.855116Z     info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
...

```

It is because client will adjust keep alive time when receive too many pings error.c And current keep alive time by default is 30s, but keepalive.EnforcementPolicy.MinTime is 1 minute.

```
// receiving a GoAway.
func (ac *addrConn) adjustParams(r transport.GoAwayReason) {
	switch r {
	case transport.GoAwayTooManyPings:
		v := 2 * ac.dopts.copts.KeepaliveParams.Time
		ac.cc.mu.Lock()
		if v > ac.cc.mkp.Time {
			ac.cc.mkp.Time = v
		}
		ac.cc.mu.Unlock()
	}
}
```

**Comment by howardjohn** [comment]
Fixed by #15254
